### DISCUSSION TITLE
symlinks, checkout, and read-only

### DISCUSSION THREAD

**Discussion by indigoviolet** [kevin-hanselman/dud#160] [General]
- If a stage has been previously committed, it is represented as symlinks to the cache
- If I later `dud checkout`, it is still symlinks to read-only files
- If I attempt to open the file for writing, this fails!
- I have to do `dud checkout --copy` to work around this.

This feels off. Should `checkout` always imply `--copy`? Clearly we don't want to be writing through to the cache.

**Comment by kevin-hanselman** [comment]
> Should `checkout` always imply `--copy`?

This is intentionally not the default behavior, because `--copy` is very slow for large files or lots of files. You can read more about why this is the default in the [README](https://github.com/kevin-hanselman/dud#dud-checks-out-files-as-symbolic-links-by-default).

> I have to do `dud checkout --copy` to work around this.

Not necessarily. To modify an artifact, I would recommend applying the following pattern in your code:

1. Read the file (really a symlink to the cache)
2. Make your changes to the file (e.g., by editing a Pandas DataFrame)
3. Delete the file (again, really a symlink)
4. Write the new file
5. In your shell, `dud commit`, etc.

Alternatively, you could use a temporary file for the new version of the artifact. That might be useful to quickly diff the artifacts.

If you could expand on your use case, I may be able to offer more/other advice.

**Comment by indigoviolet** [comment]
Thanks for the quick response!

- The case I ran into was for a log file, where I can't fully create the structure in memory before writing to disk

- I agree I can delete the symlink, but that feels like it breaks the abstraction: I feel like the pipeline code itself should be agnostic to the details of what dud is doing with the files, or even that dud exists and is orchestrating the code. So I am reluctant to abandon the `.open('w')` pattern for clobbering files and conditionally handle symlinks everywhere.

**Reply by kevin-hanselman** [reply]
I see what you mean about Dud being transparent to your code. At the same time, requiring `path.unlink()` before `path.open('w')` is a very reasonable price to pay, IMHO 😄. I'm not sure that you need to "conditionally handle symlinks everywhere".

A general philosophy I have with Dud is that your source code should be ultimately responsible for its reproducibility, not Dud. As I say in the README: "Dud's minimalist approach results in a stage's command entirely owning its own reproducibility; the responsibility is not awkwardly shared between the stage [(your code)] and the tool [(Dud)]."

In my experience, there's too many nuances and subtleties in data pipelines for a generic tool to perfectly optimize for UX and/or computational efficiency; you're better off making smart choices (e.g., work avoidance) in your code.

I'm still a little puzzled by your use case. Could you make a dummy stage file (or files) for me to get a better sense? That may help me better address your concerns.
